### PR TITLE
Staging Sites: Add staging to production sync

### PR DESCRIPTION
**Pull request by kozer**
Related to https://github.com/Automattic/dotcom-forge/issues/3692

## Proposed Changes

This pr adds a prototype for staging -> production sync

## Testing Instructions

Apply D123500-code

### Test 1: UI
1. Create a WoA site
2. Add a staging site
3. While being on the production site, ensure that you see a "Pull from staging" button like below:
2023-09-28T12:11:17,419544892+03:00

4. Press the button. Ensure that you see the following modal:
2023-09-28T12:11:53,828376063+03:00
5. Navigate to the staging side:
6. Ensure that you see a "Push to production" button.
2023-09-28T12:11:26,588236491+03:00
7. Press the button. Ensure that you see the following modal:
2023-09-28T12:11:36,956043091+03:00

### Test 2: Functionality
1. Create a WoA site
2. Add a staging site
3. Apply D123500-code.
4. At your staging site, change your theme ( eg: `Awburn` ).
5. While on the production side, press the pull button. Ensure that the endpoint called is `/wpcom/v2/sites/<prod_id>/staging-site/pull-from-staging/<staging_id>`
6. While on the staging side, press the push button. Ensure that the endpoint called is `/wpcom/v2/sites/<production_id>/staging-site/pull-from-staging/<staging_site_id>`
7. Navigate to rewind debugger. Ensure that you see a new queued job.
9. Navigate to `Appearance -> Themes` at your production site and ensure that your theme has been changed.

### Test 3: Analytics
1. Create a WoA site
2. Add a staging site
3. Apply D123500-code
4. Change the code to return an error.
5. While on the production side, press the push button. Ensure that you see `calypso_hosting_configuration_staging_site_pull_failure` event to being dispatched.
6. While on the staging side, press the pull button.  Ensure that you see `calypso_hosting_configuration_staging_site_pull_failure` being dispatched.

NOTE: Please ignore the ugly buttons for now. The UI will be redesigned in an upcoming pr

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/hosting/staging-site-card/card-content/manage-staging-site-card-content.tsx [hunk 1]
```diff
@@ -104,6 +106,24 @@ export const ManageStagingSiteCardContent = ( {
 			);
 		};
 
+		const ConfirmationPullChangesButton = () => {
+			return (
+				<ConfirmationModal
+					disabled={ isButtonDisabled }
+					onConfirm={ onPullClick }
+					modalTitle={ translate( 'Confirm pull your changes to your production site' ) }
```
Comment (wojtekn) [inline client/my-sites/hosting/staging-site-card/card-content/manage-staging-site-card-content.tsx]:
```suggestion
					modalTitle={ translate( 'Confirm pull your changes from your staging site' ) }
```

We will adjust it when we apply the design, but let's clarify those to use the same wording as in the button.

File: client/my-sites/hosting/staging-site-card/card-content/manage-staging-site-card-content.tsx [hunk 2]
```diff
@@ -104,6 +106,24 @@ export const ManageStagingSiteCardContent = ( {
...
+					modalMessage={ translate(
+						'Are you sure you want to pull your staging changes to your production site?'
```
Comment (wojtekn) [inline client/my-sites/hosting/staging-site-card/card-content/manage-staging-site-card-content.tsx]:
```suggestion
						'Are you sure you want to pull your changes from your staging site?'
```

File: client/my-sites/hosting/staging-site-card/card-content/manage-staging-site-card-content.tsx [hunk 3]
```diff
@@ -169,6 +189,7 @@ export const ManageStagingSiteCardContent = ( {
 						<LeftActionsContainer>
 							<ManageStagingSiteButton />
 							<ConfirmationPushChangesButton />
+							<ConfirmationPullChangesButton />
```
Comment (wojtekn) [inline client/my-sites/hosting/staging-site-card/card-content/manage-staging-site-card-content.tsx:192]: We will adjust those under the design ticket but for now, could we move the Push and Pull buttons to the new line under the Manage Staging Site and Delete button to improve readability?

File: client/my-sites/hosting/staging-site-card/staging-site-production-card.tsx [hunk 4]
```diff
@@ -92,18 +104,32 @@ function StagingSiteProductionCard( { disabled, siteId, translate }: CardProps )
...
-							confirmLabel={ translate( 'Pull from production' ) }
-							cancelLabel={ translate( 'Cancel' ) }
-						>
-							<Gridicon icon="arrow-down" />
-							<span>{ translate( 'Pull from production' ) }</span>
-						</ConfirmationModal>
+						<>
+							<ConfirmationModal
+								onConfirm={ pushToStaging }
+								modalTitle={ translate( 'Confirm pulling changes to your staging site.' ) }
```
Comment (wojtekn) [inline client/my-sites/hosting/staging-site-card/staging-site-production-card.tsx]:
```suggestion
								modalTitle={ translate( 'Confirm pulling changes from your production site.' ) }
```

File: client/my-sites/hosting/staging-site-card/staging-site-production-card.tsx [hunk 5]
```diff
@@ -92,18 +104,32 @@ function StagingSiteProductionCard( { disabled, siteId, translate }: CardProps )
...
+								modalMessage={ translate(
+									'Are you sure you want to pull your production changes to your staging site?'
```
Comment (wojtekn) [inline client/my-sites/hosting/staging-site-card/staging-site-production-card.tsx]:
```suggestion
									'Are you sure you want to pull your changes from your production site?'
```

File: client/my-sites/hosting/staging-site-card/staging-site-production-card.tsx [hunk 6]
```diff
@@ -92,18 +104,32 @@ function StagingSiteProductionCard( { disabled, siteId, translate }: CardProps )
...
+								cancelLabel={ translate( 'Cancel' ) }
+							>
+								<Gridicon icon="arrow-down" />
+								<span>{ translate( 'Pull from production' ) }</span>
+							</ConfirmationModal>
+							<ConfirmationModal
+								onConfirm={ pullFromStaging }
+								modalTitle={ translate( 'Confirm pushing changes to your production site.' ) }
+								modalMessage={ translate(
+									'Are you sure you want to push your staging changes to your production site?'
```
Comment (wojtekn) [inline client/my-sites/hosting/staging-site-card/staging-site-production-card.tsx]:
```suggestion
									'Are you sure you want to push your changes to your production site?'
```

File: client/my-sites/hosting/staging-site-card/card-content/manage-staging-site-card-content.tsx (hunk 3 above)
Comment (kozer) [reply client/my-sites/hosting/staging-site-card/card-content/manage-staging-site-card-content.tsx:192]: Do you think this is ok? ( I like it less than before )
2023-09-29T10:03:02,079582682+03:00

Comment (wojtekn) [reply client/my-sites/hosting/staging-site-card/card-content/manage-staging-site-card-content.tsx:192]: Yes, the previous one has all the buttons wrapped into two lines, and it feels broken.

In the final design, we will wrap the site icon, title, URL, and first two buttons in a block, and we will keep everything related to the Push/Pull actions below, so in your new screenshot we are slightly going into the desired direction.

**Review by wojtekn — APPROVED** [review]

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/9090679

Thank you @kozer for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/9090679).

**Pull request merged**